firedrake/solving_utils.py

```diff
@@ -334,8 +394,15 @@ def compute_operators(ksp, J, P):
         if fine is not None:
             _, _, inject = dmhooks.get_transfer_operators(fine._x.function_space().dm)
             inject(fine._x, ctx._x)
-            for bc in ctx._problem.bcs:
-                bc.apply(ctx._x)
+
+            def rapply(bcs, u):
+                for bc in bcs:
+                    if isinstance(bc, DirichletBC):
+                        bc.apply(u)
+                    elif isinstance(bc, EquationBC):
+                        rapply(bc.bcs, u)
```

Comment:
Can this make use of the `__iter__` methods on these objects now? If not, seems like something has gone wrong.

Reply:
I missed this one.  Now fixed!